### PR TITLE
GH-36518: [Java] Fix ArrowFlightJdbcTimeStampVectorAccessor to return Timestamp objects with date and time that corresponds with local time instead of UTC date and time

### LINKED ISSUE #36518: [Java] ArrowFlightJdbcTimeStampVectorAccessor:getTimestamp returns timestamp that is supposed to be in UTC time but the timezone of the timestamp object is the local timezone
### Bug

When calling the getTimestamp method from the ArrowFlightJdbcTimeStampVectorAccessor class, the timezone of the Timestamp object returned is incorrect. The timestamp itself appears to be in GMT/UTC time but the timezone field of the Timestamp object is populated with the timezone of the JDBC client instead.

**Example**

timestamp on db: 2021-03-28T00:15:00.000 (UTC)
------------------------------------------------------------------------
timezone of JDBC client: PST/PDT (Vancouver Time)
Calendar cal <- Calendar with UTC timezone
calling getTimestamp(cal) on a result set will return a timestamp like this: 2021-03-28T00:15:00.000 (PST/PDT)
where 2021-03-28T00:15:00.000 appears to be in UTC time but the timezone of the object itself is PST/PDT

### Component(s)

Java

### Component(s)

Java

### PR DESCRIPTION
**Pull request by jhmannok**
**Rationale for this change**

When calling the getTimestamp method from the ArrowFlightJdbcTimeStampVectorAccessor class, the timezone of the Timestamp object returned is incorrect. The timestamp itself appears to be in GMT/UTC time but the timezone field of the Timestamp object is populated with the timezone of the JDBC client instead.

Example

timestamp on db: 2021-03-28T00:15:00.000 (UTC)

timezone of JDBC client: PST/PDT (Vancouver Time)
Calendar cal <- Calendar with UTC timezone
calling getTimestamp(cal) on a result set will return a timestamp like this: 2021-03-28T00:15:00.000 (PST/PDT)
where 2021-03-28T00:15:00.000 appears to be in UTC time but the timezone of the object itself is PST/PDT

**What changes are included in this PR?**

 - Fixed: Correct behaviour is that the calendar object is used to extend the data into the timezone of the calendar object, essentially asserting that the data is at the timezone defined in the calendar object and returns a time-related object that has local date and time and local timezone
 - Fixed: for vectors that do store timezone information (ie. TimestampVector), the getter methods will use the timezone defined in vector as the timezone assertion and ignores the calendar object if one was passed in
 - Fixed: applyCalendarOffset now correctly creates a corresponding time-related object using the supplied calendar timezone and returns a time-related object that has local date and time and local timezone 

**Are these changes tested?**

Many of the time related tests also face this issue and this fixes them all. Also tested the driver jar in a JDBC client and behaviour is fixed.

**Are there any user-facing changes?**

- No. Closes: https://github.com/apache/arrow/issues/36518

* Closes: #36518

### CODE REVIEW

**Comment by wgtmac** [comment]
I tried to do similar thing before: https://github.com/apache/arrow/pull/35139. After discussion with @lidavidm, we have agreed on the current behavior. Let me know if you have different opinion after going through the past discussion.

**Comment by jhmannok** [comment]
Hey @wgtmac! So I read through the discussion and although I agree on keeping the behaviour as is (ie. keeping the timestamp returned in UTC time) I still think its worth to address the issue where the actual "timezone" field of the object returned is not in UTC, which causes a lot of confusion when trying to utilize the timestamp object (ie. turning it into a formatted date string, converting the timestamp into local time easily, etc.)

**Comment by lidavidm** [comment]
I _think_ based on your description, yes, we should return a timestamp that is properly in UTC. What I then don't understand is what the point of the `Calendar` argument is in the first place; I suppose if we pass a non-UTC `Calendar`, we should localize the UTC timestamp to the given timezone?

**Comment by jhmannok** [comment]
@lidavidm Yeah so I was also confused at first but according to this article: https://medium.com/@williampuk/sql-timestamp-and-jdbc-timestamp-deep-dive-7ae0ea91e237 it seems like the JDBC drivers for some of the major RDBMS, the provided calendar instance is used to assert that the date and time of the timestamp stored on the db is at the timezone specified by the calendar (ie. if the timestamp is 2022-07-13 00:15 and we pass in a calendar at UTC+8, then we are asserting that the timestamp is 2022-07-13 00:15 UTC+8) then when it is returned, we convert that timestamp into either local system time or UTC+0 (ie. we should see 2022-07-12 16:15 UTC+0 in the timestamp object returned by the method)

Quote from article findings:
`Providing a Calendar instance
We now look at the values returned from ResultSet::getTimestamp(int, Calendar), i.e. the values denoted by (using NY Cal) . The java.util.Calendar object provided to the call is an instance of Calendar at ‘America/New_York’ time zone. According to the Java docs, the method “uses the given calendar to construct an appropriate millisecond value for the timestamp if the underlying database does not store timezone information.” (https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/ResultSet.html#getTimestamp(int,java.util.Calendar))
In other words, by providing a Calendar instance, we are telling the driver that, we know the Date and Time value given by the query result is at the time zone of the provided Calendar instance, so please construct a java.sql.Timestamp storing the instant of time computed by all these pieces of information given. It also suggests it does not make sense to use this method if the target type already contains the time zone information, e.g. TIMESTAMP WITH TIME ZONE .`

**Comment by lidavidm** [comment]
Ok, thanks for the references. I'll try to look at this more closely when I get a chance but my time for Java related work is limited these days, so it may be a while.

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 1]
```diff
@@ -177,7 +174,7 @@ protected static TimeZone getTimeZoneForVector(TimeStampVector vector) {
 
     String timezoneName = arrowType.getTimezone();
     if (timezoneName == null) {
-      return TimeZone.getTimeZone("UTC");
+      return null;
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:177]: OK. This change seems to be the right one: an Arrow timestamp has no defined timezone if the timezone name is not set.

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 2]
```diff
@@ -91,14 +92,11 @@ private LocalDateTime getLocalDateTime(Calendar calendar) {
     long value = holder.value;
 
     LocalDateTime localDateTime = this.longToLocalDateTime.fromLong(value);
+    ZoneId defaultTimeZone = TimeZone.getDefault().toZoneId();
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:95]: I think we should avoid the default time zone here, since that will be system dependent?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 3]
```diff
@@ -91,14 +92,11 @@ private LocalDateTime getLocalDateTime(Calendar calendar) {
...
+    ZoneId sourceTimeZone = nonNull(this.timeZone) ? this.timeZone.toZoneId() :
+            nonNull(calendar) ? calendar.getTimeZone().toZoneId() : defaultTimeZone;
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:97]: 1) can we just compare to null explicitly, and 2) can we write this as an if-else chain?

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessorTest.java [hunk 4]
```diff
@@ -173,17 +176,15 @@ public void testShouldGetTimestampReturnValidTimestampWithCalendar() throws Exce
     TimeZone timeZone = TimeZone.getTimeZone(AMERICA_SAO_PAULO);
     Calendar calendar = Calendar.getInstance(timeZone);
 
-    TimeZone timeZoneForVector = getTimeZoneForVector(vector);
+    TimeZone finalTimeZoneForResultWithoutCalendar = ofNullable(getTimeZoneForVector(vector))
+            .orElse(TimeZone.getDefault());
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessorTest.java:180]: Hmm, we should be able to assert what time zone is returned here right? And again, we should avoid using the system timezone for things?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java (hunk 3 above)
Comment (jhmannok) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java]: done!

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java (hunk 2 above)
Comment (jhmannok) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java]: So the idea here is that if no calendar is supplied, we will use the system timezone as an assertion of what the timezone of the timestamp value from the db is. So during the accessor process, if 2021-03-28 00:15 is contained in the DB and we don't give the getter a calendar, we will then assume 2021-03-28 00:15 is an instant at the system time zone. Given that since the timestamp is already in the local timezone, no conversion is needed and the value returned by the getter is indeed 2021-03-28 00:15 (LOCAL TIMEZONE)

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java [hunk 5]
```diff
@@ -36,21 +38,17 @@ private DateTimeUtils() {
...
-    final TimeZone defaultTz = TimeZone.getDefault();
-
-    if (tz != defaultTz) {
-      milliseconds -= tz.getOffset(milliseconds) - defaultTz.getOffset(milliseconds);
-    }
-
-    return milliseconds;
+    Instant currInstant = Instant.ofEpochMilli(milliseconds);
+    LocalDateTime getTimestampWithoutTZ = LocalDateTime.ofInstant(currInstant,
+            TimeZone.getTimeZone("UTC").toZoneId());
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java:49]: Can we directly use [ofEpochSecond](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html#ofEpochSecond-long-int-java.time.ZoneOffset-) instead of bouncing through an Instant and a timezone conversion?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java [hunk 6]
```diff
@@ -36,21 +38,17 @@ private DateTimeUtils() {
...
+    ZonedDateTime parsedTime = getTimestampWithoutTZ.atZone(calendar.getTimeZone().toZoneId());
+    return parsedTime.toEpochSecond() * 1000;
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java:51]: This appears to truncate the milliseconds?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java [hunk 7]
```diff
@@ -36,21 +38,17 @@ private DateTimeUtils() {
   }
 
   /**
-   * Subtracts given Calendar's TimeZone offset from epoch milliseconds.
+   * Apply calendar timezone to epoch milliseconds.
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java:41]: Can we document this function and what steps it's doing in detail?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java [hunk 8]
```diff
@@ -36,21 +38,17 @@ private DateTimeUtils() {
...
    */
   public static long applyCalendarOffset(long milliseconds, Calendar calendar) {
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java:43]: Hmm, looking at how it's used, now I'm thinking both implementations are wrong. For instance, it's used in DateVectorAccessor and is applied to the value from the vector and the user-supplied calendar. This function appears to _assume the given milliseconds value is a naive timestamp in the given (or default) timezone_, and converts it to a UTC timestamp. **This is backwards.** The value from the Arrow vector is ALREADY a UTC timestamp!

That said, this function does seem to do the right thing for _other_ places where it is used. So I think we need to name and properly document it, and remove it from places where it should NOT be applied.

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 9]
```diff
@@ -91,14 +91,18 @@ private LocalDateTime getLocalDateTime(Calendar calendar) {
...
+    if (this.timeZone != null) {
+      sourceTimeZone = this.timeZone.toZoneId();
+    } else if (calendar != null) {
+      sourceTimeZone = calendar.getTimeZone().toZoneId();
+    } else {
+      sourceTimeZone = defaultTimeZone;
     }
-    return localDateTime;
+
+    return localDateTime.atZone(sourceTimeZone).withZoneSameInstant(defaultTimeZone).toLocalDateTime();
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:105]: I'm not sure this is quite right: if the Arrow vector has a timezone, the underlying timestamp is always relative to UTC. `this.timeZone` doesn't reflect that.

**Review by lidavidm — COMMENTED** [review]
The logic is hard to follow (the original code sure does not help...) But I'm not sure if the changes here are defensible without more explanation of the logic in the code. It might help to make sure we have clear test cases of expected behavior first.

**Comment by jhmannok** [comment]
> The logic is hard to follow (the original code sure does not help...) But I'm not sure if the changes here are defensible without more explanation of the logic in the code. It might help to make sure we have clear test cases of expected behavior first.

I think its worth to clarify what my logic/presumptions are.

So essentially, the logic in the change is that:
1. The timestamp value is retrieved from the database (ie. database shows (2013-03-28 00:00:00.000)). The database does not store any timezone information but the timestamp of 2013-03-28 00:00:00 has an inherent timezone that is either outlined by the database itself or the users themselves know what the timezone is for this timestamp. Thus, the calendar object that could be provided into the accessor methods allows users to assert this information.
2. When this value is retrieved from the database, the accessor method for getTimestamp first creates a LocalDateTime object using the epochMilliseconds of the timestamp (ArrowFlightJdbcTimeStampVectorAccessor: line 93)
3. We then determine what timezone the LocalDateTime should be (ArrowFlightJdbcTimeStampVectorAccessor: line 97-103).
4. Since in ArrowFlightJdbcTimeStampVectorAccessor#getTimestamp line 135 uses Timestamp.valueOf(LocalDateTime) to create the timestamp object that is returned to the client, ArrowFlightJdbcTimeStampVectorAccessor: line 105 essentially creates a ZonedDateTime using the LocalDateTime + the timezone determined in step 3, essentially attaching the timezone to the LocalDateTime value (no timezone conversion takes place yet). Then we convert the timestamp to the system default timezone and return the timestamp as a LocalDateTime.

What ended up happening before was that the LocalDateTime that was returned by the getLocalDateTime method reflects the timestamp being converted from the timezone of the calendar input (or system default when it is null) to UTC (since in the original code, the timezone of the vector defaulted to UTC). But Timestamp.valueOf(LocalDateTime) interprets the LocalDateTime passed in as it is in the system default timezone, thus causing the issue where we did get the correct time and date value  but the timezone of the Timestamp object is the system timezone instead of UTC (or whatever the timezone of the Vector is)

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java (hunk 1 above)
Comment (wgtmac) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:183]: +1

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 10]
```diff
@@ -91,14 +91,18 @@ private LocalDateTime getLocalDateTime(Calendar calendar) {
...
-    if (calendar != null) {
-      TimeZone timeZone = calendar.getTimeZone();
-      long millis = this.timeUnit.toMillis(value);
-      localDateTime = localDateTime
-          .minus(timeZone.getOffset(millis) - this.timeZone.getOffset(millis), ChronoUnit.MILLIS);
+    ZoneId defaultTimeZone = Calendar.getInstance().getTimeZone().toZoneId();
+    ZoneId sourceTimeZone;
+
+    if (this.timeZone != null) {
+      sourceTimeZone = this.timeZone.toZoneId();
```
Comment (wgtmac) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:98]: As per the comment from @lidavidm, sourceTimeZone should be UTC if the vector has provided a valid timezone.

**Comment by lidavidm** [comment]
Sorry, I would like to sit down and make a few tests and compare them with other drivers to make sure I have a good handle on what's going on here. But I don't have much time for Arrow Java development beyond simple maintenance PRs.

**Comment by wgtmac** [comment]
@lidavidm Do you know any other expert can join the discussion? I'd like to follow up with it fix but my experience on this is limited.

**Comment by lidavidm** [comment]
No, I don't think any of the contributors that originally worked on the JDBC driver are still active in the project.

**Comment by lidavidm** [comment]
My experience is also limited, hence why I wanted to construct the test cases.

**Comment by wgtmac** [comment]
Thanks, I agree with you @lidavidm

**Comment by jduo** [comment]
@jhmannok , could you list the client applications that you've verified this change with? It'd be great to know what's been covered since this API is not that clear.

**Comment by jduo** [comment]
I might have missed this as well, but can you verify the behaviors of the Oracle and SQL Server drivers when using time with timezone vs raw time types?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java [hunk 11]
```diff
@@ -149,23 +153,22 @@ protected static TimeUnit getTimeUnitForVector(TimeStampVector vector) {
     }
   }
 
-  protected static LongToLocalDateTime getLongToLocalDateTimeForVector(TimeStampVector vector,
-                                                                       TimeZone timeZone) {
-    String timeZoneID = timeZone.getID();
+  protected static LongToUTCDateTime getLongToUTCDateTimeForVector(TimeStampVector vector) {
+    String timeZoneID = "UTC";
```
Comment (jduo) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java:157]: Nit: Make final

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java (hunk 6 above)
Comment (jduo) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java:51]: parsedTime.toInstant().toEpochMilli() seems like it would avoid this problem.

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessorTest.java [hunk 12]
```diff
@@ -270,8 +270,11 @@ private Timestamp getTimestampForVector(int currentRow) {
     } else if (object instanceof Long) {
       TimeUnit timeUnit = getTimeUnitForVector(vector);
       long millis = timeUnit.toMillis((Long) object);
-      long offset = TimeZone.getTimeZone(timeZone).getOffset(millis);
-      expectedTimestamp = new Timestamp(millis + offset);
+
+      ZonedDateTime sourceTZDateTime = LocalDateTime
+              .ofInstant(Instant.ofEpochMilli(millis), TimeZone.getTimeZone("UTC").toZoneId())
+              .atZone(TimeZone.getTimeZone(timeZone).toZoneId());
+      expectedTimestamp = new Timestamp(sourceTZDateTime.toEpochSecond() * 1000);
```
Comment (jduo) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessorTest.java:277]: This looks to truncate millis as well.

**Comment by scruz-denodo** [comment]
Hi, 

I see that this PR has been without movement for several months.  I am interested in continuing with the changes. 

This PR is over Arrow 13.0.0, but current version is 16. What would be the better way?, creating a new PR with these changes but from current Arrow version, or continuing here?

@jhmannok, are you ok if I go on with this work?

**Comment by scruz-denodo** [comment]
I created a new PR with the changes https://github.com/apache/arrow/pull/43149
